### ISSUE TITLE
`self.log` raised error when number of dataloader is not consistent

### ISSUE THREAD

**Issue by ding3820** [Lightning-AI/lightning#16431]
### Bug description

Hi all,

I posted a discussion in Lightning.ai forum [here](https://lightning.ai/forums/t/self-log-raised-error-when-number-of-dataloader-is-not-consistent/2167) and @awaelchli suggested me reporting an issue. 

There might be a bug in the way `self.log` recording `dataloader_idx`. If we have two validation dataloaders, says A and B. We use A every epoch, but only use B every 2 epoch. However, while using `self.log`, in `validation_step()`, an error would show up:

```
You called self.log({name}, ...) twice in {fx} with different arguments. This is not allowed
```
(see [here](https://github.com/Lightning-AI/lightning/blob/15ef52bc732d1f907de4de58683f131652c0d68c/src/pytorch_lightning/trainer/connectors/logger_connector/result.py#L489))

The way I implement it was by switching the available dataloaders in `val_dataloader()` and reload dataloader every epoch.
```
def val_dataloader():
    if self.should_run_B():
        return [loader_A, loader_B]
    else:
        return [loader_A]
```

I notice that when we are at the epoch that only use one validation dataloader, dataloader_idx is always None. On the other hand, when we have two validation dataloaders, dataloader_idx would be sequentially presented as 0 or 1 in validation_step. If I understand correctly, this is the main reason causing the error.

Another interesting finding is that if we set `add_dataloader_idx=True` for all `self.log`, the program would run without error. But the tensorboard logging would be wrong that shows and `c_0` and`c_0/dataloader_idx_0` together (See snapshot below). These two were meant to remain in the same figure but somehow it got split into two figures. Probably it is because of the alternating dataloader design.
<img width="865" alt="image" src="https://user-images.githubusercontent.com/28510501/213349188-f087d23d-c431-4ac6-b0cf-323ae996be15.png">

### How to reproduce the bug

```python
import torch
from torch.utils.data import DataLoader
from pytorch_lightning.demos.boring_classes import BoringModel, RandomDataset
from pytorch_lightning import Trainer

class TestModel(BoringModel):
    def training_step(self, batch, batch_idx):
        out = super().training_step(batch, batch_idx)
        self.log("a", out["loss"])
        self.log("b", out["loss"], on_step=True, on_epoch=True)
        return out

    def validation_step(self, batch, batch_idx, dataloader_idx=0):
        out = super().validation_step(batch, batch_idx)
        if dataloader_idx == 0:
            self.log("c_0", out["x"], add_dataloader_idx=False)
            self.log("d_0", out["x"], on_step=True, on_epoch=True, add_dataloader_idx=False)
        elif dataloader_idx == 1:
            self.log("c_1", out["x"], add_dataloader_idx=False)
            self.log("d_1", out["x"], on_step=True, on_epoch=True, add_dataloader_idx=False)
        return out

    def validation_epoch_end(self, outputs):
        self.log("g", torch.tensor(2, device=self.device), on_epoch=True)

    def val_dataloader(self):
        if self.current_epoch % 2:
            return [DataLoader(RandomDataset(32, 64)), DataLoader(RandomDataset(32, 64))]
        else:
            return DataLoader(RandomDataset(32, 64))

model = TestModel()

trainer = Trainer(
    reload_dataloaders_every_n_epochs=1,
    default_root_dir="test",
    max_epochs=10,
    log_every_n_steps=1,
    enable_model_summary=False,
)
trainer.fit(model)
```

### Error messages and logs

```
You called `self.log(c_0, ...)` twice in `validation_step` with different arguments. This is not allowed
```

### Environment

* CUDA:
	- GPU:
		- NVIDIA TITAN RTX
	- available:         True
	- version:           11.7
* Lightning:
	- pytorch-lightning: 1.7.2
	- pytorch-quantization: 2.1.2
	- torch:             1.12.0a0+8a1a93a
	- torch-tensorrt:    1.1.0a0
	- torchmetrics:      0.9.3
	- torchtext:         0.13.0a0
	- torchvision:       0.13.0a0
* Packages:
	- absl-py:           1.0.0
	- acsconv:           0.1.1
	- aiohttp:           3.8.1
	- aiosignal:         1.2.0
	- alabaster:         0.7.12
	- ansible:           6.7.0
	- ansible-core:      2.13.7
	- ansible-vault:     2.1.0
	- apex:              0.1
	- appdirs:           1.4.4
	- argon2-cffi:       21.3.0
	- argon2-cffi-bindings: 21.2.0
	- argparse:          1.4.0
	- asttokens:         2.0.5
	- async-timeout:     4.0.2
	- attrs:             21.4.0
	- audioread:         2.1.9
	- autopep8:          1.7.0
	- babel:             2.10.1
	- backcall:          0.2.0
	- backports.functools-lru-cache: 1.6.4
	- batchgenerators:   0.24
	- beautifulsoup4:    4.11.1
	- black:             22.6.0
	- bleach:            5.0.0
	- blis:              0.7.7
	- brotlipy:          0.7.0
	- cachetools:        5.0.0
	- catalogue:         2.0.6
	- certifi:           2021.10.8
	- cffi:              1.15.0
	- cfgv:              3.3.1
	- chardet:           4.0.0
	- charset-normalizer: 2.0.12
	- click:             8.0.4
	- cloudpickle:       2.0.0
	- codecov:           2.1.12
	- colorama:          0.4.4
	- conda:             4.12.0
	- conda-build:       3.21.8
	- conda-package-handling: 1.8.1
	- coverage:          6.3.2
	- cryptography:      36.0.2
	- cuda-python:       11.6.1
	- cudf:              22.4.0a0+306.g0cb75a4913
	- cugraph:           22.4.0a0+102.g4106a188
	- cuml:              22.4.0a0+108.g2be11269d
	- cupy-cuda115:      9.6.0
	- cycler:            0.11.0
	- cymem:             2.0.6
	- cython:            0.29.28
	- dask:              2022.3.0
	- dask-cuda:         22.4.0
	- dask-cudf:         22.4.0a0+306.g0cb75a4913
	- dataclasses:       0.8
	- debugpy:           1.6.0
	- decorator:         5.1.1
	- defusedxml:        0.7.1
	- dicompyler-core:   0.5.5
	- distlib:           0.3.6
	- distributed:       2022.3.0
	- docopt:            0.6.2
	- docutils:          0.17.1
	- einops:            0.4.1
	- entrypoints:       0.4
	- executing:         0.8.3
	- expecttest:        0.1.3
	- fastjsonschema:    2.15.3
	- fastrlock:         0.8
	- filelock:          3.6.0
	- fire:              0.4.0
	- flake8:            5.0.4
	- flask:             2.1.2
	- fonttools:         4.33.3
	- frozenlist:        1.3.1
	- fsspec:            2022.3.0
	- future:            0.18.2
	- glob2:             0.7
	- gnureadline:       8.1.2
	- google-auth:       2.6.6
	- google-auth-oauthlib: 0.4.6
	- graphsurgeon:      0.4.5
	- grpcio:            1.46.0
	- grpcio-tools:      1.46.0
	- h5py:              3.7.0
	- heapdict:          1.0.1
	- hypothesis:        4.50.8
	- identify:          2.5.9
	- idna:              3.3
	- imageio:           2.22.4
	- imagesize:         1.3.0
	- importlib-metadata: 4.11.3
	- importlib-resources: 5.7.1
	- iniconfig:         1.1.1
	- ipykernel:         6.13.0
	- ipython:           8.3.0
	- ipython-genutils:  0.2.0
	- isort:             5.10.1
	- itsdangerous:      2.1.2
	- jedi:              0.18.1
	- jinja2:            3.1.2
	- joblib:            1.1.0
	- json5:             0.9.6
	- jsonschema:        4.5.1
	- jupyter-client:    7.3.0
	- jupyter-core:      4.10.0
	- jupyter-tensorboard: 0.2.0
	- jupyterlab:        2.3.2
	- jupyterlab-pygments: 0.2.2
	- jupyterlab-server: 1.2.0
	- jupytext:          1.13.8
	- kiwisolver:        1.4.2
	- kornia:            0.6.6
	- langcodes:         3.3.0
	- libarchive-c:      4.0
	- librosa:           0.8.1
	- linecache2:        1.0.0
	- llvmlite:          0.36.0
	- lmdb:              1.3.0
	- locket:            1.0.0
	- markdown:          3.3.7
	- markdown-it-py:    2.1.0
	- markupsafe:        2.1.1
	- matplotlib:        3.5.2
	- matplotlib-inline: 0.1.3
	- mccabe:            0.7.0
	- mdit-py-plugins:   0.3.0
	- mdurl:             0.1.1
	- medpy:             0.4.0
	- mistune:           0.8.4
	- mock:              4.0.3
	- monai:             0.9.1
	- msgpack:           1.0.3
	- multidict:         6.0.2
	- murmurhash:        1.0.7
	- mypy:              0.991
	- mypy-extensions:   0.4.3
	- nbclient:          0.6.2
	- nbconvert:         6.5.0
	- nbformat:          5.4.0
	- nest-asyncio:      1.5.5
	- networkx:          2.6.3
	- nibabel:           4.0.1
	- nltk:              3.7
	- nodeenv:           1.7.0
	- notebook:          6.4.1
	- numba:             0.53.1
	- numpy:             1.22.3
	- nvidia-dali-cuda110: 1.13.0
	- nvidia-pyindex:    1.0.9
	- nvtx:              0.2.4
	- oauthlib:          3.2.0
	- onnx:              1.11.0
	- opencv-python-headless: 4.5.5.64
	- openslide-python:  1.2.0
	- packaging:         21.3
	- pandas:            1.3.5
	- pandocfilters:     1.5.0
	- parameterized:     0.8.1
	- parso:             0.8.3
	- partd:             1.2.0
	- pathspec:          0.10.3
	- pathy:             0.6.1
	- pexpect:           4.8.0
	- pickleshare:       0.7.5
	- pillow:            9.0.1
	- pip:               22.2.2
	- pkginfo:           1.8.2
	- platformdirs:      2.6.0
	- pluggy:            1.0.0
	- polygraphy:        0.33.0
	- pooch:             1.6.0
	- pre-commit:        2.20.0
	- pre-commit-hooks:  4.3.0
	- preshed:           3.0.6
	- prettytable:       3.3.0
	- prometheus-client: 0.14.1
	- prompt-toolkit:    3.0.29
	- protobuf:          3.19.4
	- psutil:            5.9.0
	- ptyprocess:        0.7.0
	- pure-eval:         0.2.2
	- py:                1.11.0
	- pyarmor:           7.6.1
	- pyarrow:           6.0.1
	- pyasn1:            0.4.8
	- pyasn1-modules:    0.2.8
	- pybind11:          2.9.2
	- pycocotools:       2.0+nv0.6.0
	- pycodestyle:       2.9.1
	- pycosat:           0.6.3
	- pycparser:         2.21
	- pydantic:          1.8.2
	- pydeprecate:       0.3.2
	- pydicom:           2.3.0
	- pydot:             1.4.2
	- pyflakes:          2.5.0
	- pygments:          2.12.0
	- pykwalify:         1.8.0
	- pynvml:            11.4.1
	- pyopenssl:         22.0.0
	- pyparsing:         3.0.8
	- pyradiomics:       3.0.1
	- pyrsistent:        0.18.1
	- pysocks:           1.7.1
	- pytest:            6.2.5
	- pytest-cov:        3.0.0
	- pytest-pythonpath: 0.7.4
	- python-dateutil:   2.8.2
	- python-gdcm:       3.0.15
	- python-hostlist:   1.21
	- python-magic:      0.4.27
	- python-nvd3:       0.15.0
	- python-slugify:    6.1.2
	- pytorch-lightning: 1.7.2
	- pytorch-quantization: 2.1.2
	- pytz:              2022.1
	- pywavelets:        1.4.1
	- pyyaml:            6.0
	- pyzmq:             22.3.0
	- raft:              22.4.0a0+113.gf5d2627
	- regex:             2022.4.24
	- requests:          2.27.1
	- requests-oauthlib: 1.3.1
	- resampy:           0.2.2
	- resolvelib:        0.8.1
	- revtok:            0.0.3
	- rmm:               22.4.0a0+50.gf82d458
	- rsa:               4.8
	- ruamel-yaml-conda: 0.15.80
	- ruamel.yaml:       0.17.21
	- ruamel.yaml.clib:  0.2.7
	- sacremoses:        0.0.53
	- scikit-image:      0.19.3
	- scikit-learn:      1.1.2
	- scipy:             1.9.2
	- send2trash:        1.8.0
	- setuptools:        59.5.0
	- shapely:           1.8.4
	- shellingham:       1.4.0
	- simpleitk:         2.1.1.2
	- six:               1.16.0
	- smart-open:        5.2.1
	- snowballstemmer:   2.2.0
	- sortedcontainers:  2.4.0
	- soundfile:         0.10.3.post1
	- soupsieve:         2.3.1
	- spacy:             3.3.0
	- spacy-legacy:      3.0.9
	- spacy-loggers:     1.0.2
	- sphinx:            4.5.0
	- sphinx-glpi-theme: 0.3
	- sphinx-rtd-theme:  1.0.0
	- sphinxcontrib-applehelp: 1.0.2
	- sphinxcontrib-devhelp: 1.0.2
	- sphinxcontrib-htmlhelp: 2.0.0
	- sphinxcontrib-jsmath: 1.0.1
	- sphinxcontrib-qthelp: 1.0.3
	- sphinxcontrib-serializinghtml: 1.1.5
	- srsly:             2.4.3
	- stack-data:        0.2.0
	- tabulate:          0.8.9
	- tblib:             1.7.0
	- tensorboard:       2.10.0
	- tensorboard-data-server: 0.6.1
	- tensorboard-plugin-wit: 1.8.1
	- tensorboardx:      2.5.1
	- tensorrt:          8.2.5.1
	- termcolor:         1.1.0
	- terminado:         0.13.3
	- text-unidecode:    1.3
	- thinc:             8.0.15
	- threadpoolctl:     3.1.0
	- tifffile:          2022.10.10
	- timm:              0.6.7
	- tinycss2:          1.1.1
	- toml:              0.10.2
	- tomli:             2.0.1
	- toolz:             0.11.2
	- torch:             1.12.0a0+8a1a93a
	- torch-tensorrt:    1.1.0a0
	- torchmetrics:      0.9.3
	- torchtext:         0.13.0a0
	- torchvision:       0.13.0a0
	- tornado:           6.1
	- tqdm:              4.64.0
	- traceback2:        1.4.0
	- traitlets:         5.1.1
	- treelite:          2.3.0
	- treelite-runtime:  2.3.0
	- typer:             0.4.1
	- typing-extensions: 4.2.0
	- ucx-py:            0.25.0a0+13.ga16f8a2
	- uff:               0.6.9
	- unittest2:         1.1.0
	- urllib3:           1.26.9
	- virtualenv:        20.17.1
	- waitress:          2.1.2
	- wasabi:            0.9.1
	- wcwidth:           0.2.5
	- webencodings:      0.5.1
	- websocket-client:  1.3.3
	- werkzeug:          2.1.2
	- wheel:             0.37.1
	- xgboost:           1.5.2
	- yarl:              1.8.1
	- zict:              2.2.0
	- zipp:              3.8.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.8.13
	- version:           #63-Ubuntu SMP Thu Nov 24 13:43:17 UTC 2022

**Comment by jin1041** [comment]
Have you solved the problem? I met this question too.

**Comment by Pdevadiga45** [comment]
Hi, I reproduced this on current master. _ResultCollection keeps logging metadata across dataloader reloads, so changing between one and multiple validation loaders changes dataloader_idx for the same metric and triggers the mismatch error.

I have a local fix with regression tests. Is this still open for a PR?
